server/src/transport/ws.rs

```diff
@@ -334,3 +329,120 @@ async fn graceful_shutdown<S>(
+///     svc: ServiceData,
+///     rpc_service: RpcServiceBuilder<L>,
+///     mut disconnect: tokio::sync::mpsc::Receiver<()>
+/// ) -> hyper::Response<hyper::Body>
+/// where
+///     L: for<'a> tower::Layer<RpcService> + 'static,
+///     <L as tower::Layer<RpcService>>::Service: Send + Sync + 'static,
+///     for<'a> <L as tower::Layer<RpcService>>::Service: RpcServiceT<'a> + 'static,
+/// {
+///   match ws::connect(req, svc, rpc_service).await {
+///     Ok((rp, conn_fut)) => {
+///         tokio::spawn(async move {
+///             // Keep the connection alive until
+///             // a close signal is sent.
+///             tokio::select! {
+///                 _ = conn_fut => (),
+///                 _ = disconnect.recv() => (),
+///             }
+///         });
+///         rp
+///     }
+///     Err(rp) => rp,
+///   }
+/// }
+/// ```
+pub async fn connect<L>(
+	req: hyper::Request<hyper::Body>,
+	params: ServiceData,
+	rpc_middleware: RpcServiceBuilder<L>,
+) -> Result<(hyper::Response<hyper::Body>, impl Future<Output = ()>), hyper::Response<hyper::Body>>
+where
+	L: for<'a> tower::Layer<RpcService>,
+	<L as tower::Layer<RpcService>>::Service: Send + Sync + 'static,
+	for<'a> <L as tower::Layer<RpcService>>::Service: RpcServiceT<'a>,
+{
+	let mut server = soketto::handshake::http::Server::new();
+
+	match server.receive_request(&req) {
+		Ok(response) => {
+			let (tx, rx) = mpsc::channel::<String>(params.cfg.message_buffer_capacity as usize);
+			let sink = MethodSink::new(tx);
+
+			// On each method call the `pending_calls` is cloned
+			// then when all pending_calls are dropped
+			// a graceful shutdown can has occur.
+			let (pending_calls, pending_calls_completed) = mpsc::channel::<()>(1);
+
+			let rpc_service_cfg = RpcServiceCfg::CallsAndSubscriptions {
+				bounded_subscriptions: BoundedSubscriptions::new(params.cfg.max_subscriptions_per_connection),
+				id_provider: params.cfg.id_provider.clone(),
+				sink: sink.clone(),
+				_pending_calls: pending_calls,
+			};
+
+			let rpc_service = RpcService::new(
+				params.methods.clone(),
+				params.cfg.max_response_body_size as usize,
+				params.conn_id as usize,
+				rpc_service_cfg,
+			);
+
+			let rpc_service = rpc_middleware.service(rpc_service);
+
+			let fut = async move {
+				let upgraded = match hyper::upgrade::on(req).await {
+					Ok(u) => u,
+					Err(e) => {
+						tracing::debug!("Could not upgrade connection: {}", e);
+						return;
+					}
+				};
+
+				let stream = BufReader::new(BufWriter::new(upgraded.compat()));
+				let mut ws_builder = server.into_builder(stream);
+				ws_builder.set_max_message_size(params.cfg.max_response_body_size as usize);
+				let (sender, receiver) = ws_builder.finish();
+
+				let params = BackgroundTaskParams {
+					other: params,
+					ws_sender: sender,
+					ws_receiver: receiver,
+					rpc_service,
+					sink,
+					rx,
+					pending_calls_completed,
+				};
+
+				background_task(params).await;
```

Comment:
Perhaps we should improve this to send a message to background task instead of just dropping the future.

Reply:
How do you mean? I don't quite follow and am probably being dumb :)

Reply:
I was little scared about cancel-safety and it may be better to inject a channel in the background task and send a message/drop the channel to have clean shutdown.

However, I think it should be safe to drop because it's just to soketto stuff that isn't cancel-safe and that is wrapped in stream anyway.